### ISSUE TITLE
NoShadowedVariable fires on named function expression

### ISSUE THREAD

**Issue by magul** [palantir/tslint#3074]
### Bug Report

- __TSLint version__: 5.5.0 (introduced in https://github.com/palantir/tslint/commit/08ccd4778f9288e1cb5c4433c9ed51a35e1ea04f)
- __TypeScript version__: 2.4.1
- __Running TSLint via__: (pick one) grunt-tslint/node

#### TypeScript code being linted

```ts
const callbackFuntion = function callbackFunction() {};
```

with `tslint.json` configuration:

```json
{
  "defaultSeverity": "error",
  "extends": [
    "tslint:recommended"
  ],
  "jsRules": {},
  "rules": {
    "quotemark": [
      "single",
      "avoid-escape"
    ],
    "only-arrow-functions": false,
    "trailing-comma": false,
    "variable-name": [
      "allow-leading-underscore",
      "allow-trailing-underscore",
      "allow-pascal-case",
      "ban-keywords",
      "check-format"
    ],
    "one-variable-per-declaration": false,
    "no-reference": false,
    "member-access": false,
    "object-literal-sort-keys": [true, "ignore-case"]
  },
  "rulesDirectory": []
}

```

#### Actual behavior
```
ERROR: callbackFunction.ts[1, 22]: Shadowed name: 'callbackFunction'
```
#### Expected behavior
tslint pass without error

**Comment by ajafff** [comment]
IMO that error is expected. Inside the function `callbackFunction` may be something else than outside of the function. I know you are using `const` here, but you can still do something like this:
```ts
const callbackFunction = condition
    ? function callbackFunction() {doStuff(callbackFunction)}
    : function () {doOtherStuff()};
doStuff(callbackFunction);
```
Would you expect that both calls to `doStuff` could use a different function as argument?

There are several ways to prevent this error in your code:
* use a different name for the function and the variable it's assigned to (probably not preferred)
* make it an anonymous function expression by removing the name
* convert it to a function declaration
* use an arrow function

**Comment by magul** [comment]
ACK

I just asked because it's new feature from June and e.g. in eslint such syntax doesn't raise an error.

Feel free to close it, or maybe we will wait until more members of community will give his/her opinions.

**Comment by ajafff** [comment]
> maybe we will wait until more members of community will give his/her opinions.

Agreed. We should await more feedback.

**Comment by adidahiya** [comment]
Is there any advantage to

```ts
const callbackFuntion = function callbackFunction() { ... };
```

over simply

```ts
function callbackFunction() {
    ...
}
```

?

**Comment by lddubeau** [comment]
@adidahiya I've used the form `const callbackFuntion = function callbackFunction() { ... };` to work around the compiler error:

> Function declarations are not allowed inside blocks in strict mode when targeting 'ES3' or 'ES5'. Class definitions are automatically in strict mode.

Without the `const ... =` the resulting ES5 or ES3 code might be something like:

```
if (condition) {
  function callbackFunction() {}
  something(callbackFunction);
}
```

which is exactly what the error message is yelling against due to inconsistencies in how the function definition in the `if` block is going to be interpreted.

With the workaround, it is something like:

```
if (condition) {
  var callbackFunction = function callbackFunction() {}
  something(callbackFunction);
}
```

The 2nd case has none of the [inconsistencies](http://whereswalden.com/2011/01/24/new-es5-strict-mode-requirement-function-statements-not-at-top-level-of-a-program-or-function-are-prohibited/) caused by defining functions elsewhere than in the top scope or directly in a function body.

I would use an arrow function if I could but in the cases I'm talking about my function needs to have its own value for `this`.

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
